metagraph/src/annotation/taxonomy/tax_classifier.hpp

```diff
@@ -0,0 +1,201 @@
+    TaxonomyBase(double lca_coverage_rate, double kmers_discovery_rate)
+        : lca_coverage_rate_(lca_coverage_rate),
+          kmers_discovery_rate_(kmers_discovery_rate) {}
+    virtual ~TaxonomyBase() {}
+
+    TaxId assign_class(const std::string &sequence) const;
+
+  protected:
+    virtual TaxId find_lca(const std::vector<TaxId> &taxids) const = 0;
+
+    std::string get_accession_version_from_label(const std::string &label) const;
+
+    bool get_taxid_from_label(const std::string &label, TaxId *taxid) const;
+
+    /** Reads the accession version to taxid lookup table.
+     * If 'anno_matrix' is mentioned, only the labels that exist in the given annotation matrix will be stored.
+     * If 'anno_matrix' is not mentioned, the entire content of 'filepath' will be read and stored.
+     *
+     * @param [input] filepath -> a ".accession2taxid" file.
+     * @param [optional input] anno_matrix -> pointer to the annotation matrix.
+     */
+    void read_accversion_to_taxid_map(const std::string &filepath, const graph::AnnotatedDBG *anno_matrix);
+
+    /**
+     * Update the current node_scores and best_lca by taking into account the weight of the start_node and
+     * all its ancestors.
+     *
+     * @param [input] 'start_node' -> the starting node to update 'node_scores'.
+     * @param [input] 'num_kmers_per_node[taxid]' -> the number of kmers 'k' with taxonomic_map[k]=taxid.
+     * @param [input] 'desired_number_kmers' -> the threshold score that a node has to exceed in order to be
+     *                                          considered as a valid solution.
+     * @param [modified] 'node_scores' -> the current score for each node in the tree.
+     * @param [modified] 'nodes_already_propagated' -> the set of nodes that were previously processed.
+     * @param [modified] 'best_lca' -> the current classification prediction (node that exceeds the
+     *                                 `desired_number_kmers` threshold and is placed as close as possible to the leaves).
+     * @param [modified] 'best_lca_dist_to_root' -> the distance to the root for the current classification prediction.
+     */
+    void update_scores_and_lca(TaxId start_node,
+                               const tsl::hopscotch_map<TaxId, uint64_t> &num_kmers_per_node,
+                               uint64_t desired_number_kmers,
+                               tsl::hopscotch_map<TaxId, uint64_t> *node_scores,
+                               tsl::hopscotch_set<TaxId> *nodes_already_propagated,
+                               TaxId *best_lca,
+                               uint32_t *best_lca_dist_to_root) const;
+
+    /**
+     * Get the list of LCA taxid for each kmer in a given sequences.
+     * The sequence can be given in forward or in reversed orientation.
+     */
+    virtual std::vector<TaxId> get_lca_taxids_for_seq(const std::string_view &sequence, bool reversed) const = 0;
+
+    LabelType label_type_ = LabelType::UNASSIGNED;
```

Comment:
Does it make sense to allow it to be unassigned?

Reply:
you are right. Removed now